### PR TITLE
Fixes #393 event racing

### LINKED ISSUE #393: Update territories while loading in changes can cause inconsistencies in client.

### PR DESCRIPTION
**Pull request by kecso**
Now the pulling (the actual switch between the next and the current state of the project) will wait for the updateTerritory requests to finish. Parallely it loads everything, so the actual switch can happen in a single synchronous step and we should not loose any event.

### CODE REVIEW

File: src/client/js/client.js [hunk 1]
```diff
@@ -1409,67 +1303,211 @@ define([
         }
 
         function _updateTerritoryAllDone(guid, patterns, error) {
+            refreshMetaNodes(state.nodes, state.nodes);
+
             if (state.users[guid]) {
-                state.users[guid].PATTERNS = JSON.parse(JSON.stringify(patterns));
+                state.users[guid].PATTERNS = COPY(patterns);
```
Comment (pmeijer) [inline src/client/js/client.js]: Where is COPY defined?

File: src/client/js/client.js [hunk 2]
```diff
@@ -1409,67 +1303,211 @@ define([
...
+
+            if (state.ongoingLoadPatternsCounter !== 0) {
+                logger.error('at the start of loading the counter should be zero!!! [' + state.ongoingTerritoryUpdateCounter + ']');
+            }
+
+            state.loadingStatus = null;
+            state.loadNodes = {};
+            state.loading.rootHash = newRootHash;
+            state.loading.commitHash = newCommitHash;
+            state.loading.next = callback || function () {
```
Comment (pmeijer) [inline src/client/js/client.js]: Doesn't loading require a callback?

File: src/client/js/client.js [hunk 3]
```diff
@@ -1409,67 +1303,211 @@ define([
...
+        }
+
+        function loading(newRootHash, newCommitHash, callback) {
+            var i, j,
+                userIds,
+                patternPaths,
+                patternsToLoad = [];
+
+            if (state.ongoingLoadPatternsCounter !== 0) {
+                logger.error('at the start of loading the counter should be zero!!! [' + state.ongoingTerritoryUpdateCounter + ']');
```
Comment (pmeijer) [inline src/client/js/client.js]: If we're f-ed here, why not throw an Exception with the same error message. Some stack-trace couldn't heart..

File: src/client/js/client.js [hunk 4]
```diff
@@ -1409,67 +1303,211 @@ define([
...
+                state.metaNodes[newPaths[i]] = newSource[newPaths[i]].node;
+            }
+
+            for (i = 0; i < pathsToRemove.length; i += 1) {
+                delete state.metaNodes[pathsToRemove[i]];
+            }
+        }
+
+        function switchStates() {
+            //it is safe now to move the loadNodes into nodes,
```
Comment (pmeijer) [inline src/client/js/client.js]: I would add more logging in general, but at least a debug message that we're switching state here.

File: src/client/js/client.js [hunk 5]
```diff
@@ -63,11 +63,24 @@ define([
...
+                ongoingLoadPatternsCounter: 0,
+                pendingTerritoryUpdatePatterns: {},
+                loadingStatus: null,
+                inLoading: false,
+                loading: {
+                    rootHash: null,
+                    commitHash: null,
+                    next: null
+                }
+
```
Comment (pmeijer) [inline src/client/js/client.js]: Could these be grouped (together with users, nodes and loadNodes etc..)?

**Comment by pmeijer** [comment]
Tamas were you sure that the new state variables shouldn't be reintialized when switching project or branch?
Can you also try to run it with debug = true? Afraid the newly introduced keys might not be stringified..

**Comment by kecso** [comment]
Patrik, I've checked the situations regarding branch and project changes. In both scenarios, it is safer to wait for the updateTerritory requests to be resolved before switching to a new state (whether it is a new branch or project). This means, that if we would clear out any of the counters, than we could actually cause errors later (as there is a chance that some UT requests are ongoing at the time of switch).

**Pull request merged**